File: bubbler_app/lib/data/models/topics.dart
/// Curated topic list — mirrors `backend/app/db/topics.py` `KNOWN_TOPICS`.
abstract final class KnownTopics {
  static const String defaultTopic = 'general';

  static const List<String> all = [
    defaultTopic,
    'politics',
    'technology',
    'science',
    'entertainment',
    'sports',
    'business',
    'health',
    'education',
    'environment',
  ];

  static String displayName(String topic) {
    if (topic.isEmpty) return topic;
    return topic[0].toUpperCase() + topic.substring(1);
  }

  /// Returns the canonical known topic for [value], or `null` if not in [all].
  static String? resolve(String value) {
    final trimmed = value.trim();
    if (trimmed.isEmpty) return null;
    final lower = trimmed.toLowerCase();
    for (final topic in all) {
      if (topic.toLowerCase() == lower) return topic;
    }
    return null;
  }

  /// Topics whose names contain [query] (case-insensitive), excluding
  /// already-selected ones.
  static List<String> matching(
    String query, {
    List<String> excluding = const [],
  }) {
    final trimmed = query.trim();
    final excluded = excluding.map((e) => e.toLowerCase()).toSet();
    final available =
        all.where((t) => !excluded.contains(t.toLowerCase())).toList();

    if (trimmed.isEmpty) return available;

    final needle = trimmed.toLowerCase();
    return available.where((t) => t.toLowerCase().contains(needle)).toList();
  }
}

/// Helpers for cleaning and mutating topic preference lists.
abstract final class TopicPreferenceList {
  static List<String> cleaned(List<String> topics) {
    final seen = <String>{};
    final result = <String>[];

    for (final raw in topics) {
      final topic = normalizedTopic(raw);
      if (topic.isEmpty) continue;
      if (!seen.add(topic.toLowerCase())) continue;
      result.add(topic);
    }

    result.sort((a, b) => a.toLowerCase().compareTo(b.toLowerCase()));
    return result;
  }

  static List<String> add(String rawTopic, List<String> topics) {
    final topic = KnownTopics.resolve(rawTopic);
    if (topic == null) return cleaned(topics);

    final updated = cleaned(topics);
    if (updated.any((t) => t.toLowerCase() == topic.toLowerCase())) {
      return updated;
    }
    updated.add(topic);
    return cleaned(updated);
  }

  static List<String> remove(String rawTopic, List<String> topics) {
    final topic = normalizedTopic(rawTopic);
    return cleaned(topics)
        .where((t) => t.toLowerCase() != topic.toLowerCase())
        .toList();
  }

  static String normalizedTopic(String value) {
    return value.trim().replaceAll(',', '');
  }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/model/Topics.kt
package com.bubbler.android.data.model

/**
 * Curated topic list — mirrors `backend/app/db/topics.py` KNOWN_TOPICS —
 * plus preference-list clean/add/remove helpers (Swift `TopicPreferenceList`).
 */
object KnownTopics {
    const val DEFAULT_TOPIC = "general"

    val ALL: List<String> = listOf(
        DEFAULT_TOPIC,
        "politics",
        "technology",
        "science",
        "entertainment",
        "sports",
        "business",
        "health",
        "education",
        "environment",
    )

    fun displayName(topic: String): String {
        if (topic.isEmpty()) return topic
        return topic.replaceFirstChar { it.titlecase() }
    }

    /** Returns the canonical known topic for [value], or null if not in [ALL]. */
    fun resolve(value: String): String? {
        val trimmed = value.trim()
        if (trimmed.isEmpty()) return null
        return ALL.firstOrNull { it.equals(trimmed, ignoreCase = true) }
    }

    /**
     * Topics whose names contain [query] (case-insensitive),
     * excluding already-selected ones.
     */
    fun matching(query: String, excluding: List<String> = emptyList()): List<String> {
        val trimmed = query.trim()
        val excluded = excluding.map { it.lowercase() }.toSet()
        val available = ALL.filter { it.lowercase() !in excluded }
        if (trimmed.isEmpty()) return available
        return available.filter { it.contains(trimmed, ignoreCase = true) }
    }
}

object TopicPreferenceList {
    fun cleaned(topics: List<String>): List<String> {
        val seen = mutableSetOf<String>()
        return topics
            .map { normalizedTopic(it) }
            .filter { it.isNotEmpty() }
            .filter { topic -> seen.add(topic.lowercase()) }
            .sortedWith(String.CASE_INSENSITIVE_ORDER)
    }

    fun add(rawTopic: String, to: List<String>): List<String> {
        val topic = KnownTopics.resolve(rawTopic) ?: return cleaned(to)
        val updated = cleaned(to).toMutableList()
        if (updated.any { it.equals(topic, ignoreCase = true) }) {
            return updated
        }
        updated.add(topic)
        return cleaned(updated)
    }

    fun remove(rawTopic: String, from: List<String>): List<String> {
        val topic = normalizedTopic(rawTopic)
        return cleaned(from).filter { !it.equals(topic, ignoreCase = true) }
    }

    fun normalizedTopic(value: String): String =
        value.trim().replace(",", "")
}
